File: 07-数据存储/04-搜索引擎/模糊查询与编辑距离自动机/go/fuzzy.go
// FuzzyQuery 参数语义与 Levenshtein 自动机的参数化描述（Go 侧等价实现）。
//
// 转写自 apache/lucene@main
//   lucene/core/src/java/org/apache/lucene/search/FuzzyQuery.java
//   lucene/core/src/java/org/apache/lucene/util/automaton/LevenshteinAutomata.java
//   .../automaton/Lev1ParametricDescription.java / Lev2ParametricDescription.java /
//   Lev1TParametricDescription.java
//
// 语言差异（已在代码中显式落地）：
//   * Java 抛 IllegalArgumentException，Go 侧用 error 返回；
//   * Java 的 (int) 强转是向零截断，Go 的 int(float64) 同样是向零截断，语义一致；
//   * Java 的 assert 在 Go 里用 panic 对应。
package main

import (
	"errors"
	"fmt"
	"sort"
)

// MaximumSupportedDistance LevenshteinAutomata.MAXIMUM_SUPPORTED_DISTANCE
const MaximumSupportedDistance = 2

// CharacterMaxCodePoint Character.MAX_CODE_POINT
const CharacterMaxCodePoint = 0x10FFFF

// ------------------------------------------------------------------ FuzzyQuery

// FuzzyQuery 构造参数与校验（源码四条 throw 全部复刻）。
type FuzzyQuery struct {
	Term           string
	MaxEdits       int
	PrefixLength   int
	MaxExpansions  int
	Transpositions bool
}

// FuzzyQuery 的默认值
const (
	DefaultMaxEdits         = MaximumSupportedDistance // = 2
	DefaultPrefixLength     = 0
	DefaultMaxExpansions    = 50
	DefaultTranspositions   = true
)

// NewFuzzyQuery 参数校验顺序与源码一致：maxEdits → prefixLength → maxExpansions。
func NewFuzzyQuery(term string, maxEdits, prefixLength, maxExpansions int,
	transpositions bool) (*FuzzyQuery, error) {
	if maxEdits < 0 || maxEdits > MaximumSupportedDistance {
		return nil, fmt.Errorf("maxEdits must be between 0 and %d", MaximumSupportedDistance)
	}
	if prefixLength < 0 {
		return nil, errors.New("prefixLength cannot be negative.")
	}
	if maxExpansions <= 0 {
		return nil, errors.New("maxExpansions must be positive.")
	}
	return &FuzzyQuery{term, maxEdits, prefixLength, maxExpansions, transpositions}, nil
}

// TermsEnumKind maxEdits == 0 时走 SingleTermsEnum（只能精确匹配），否则 FuzzyTermsEnum。
func (q *FuzzyQuery) TermsEnumKind() string {
	if q.MaxEdits == 0 {
		return "SingleTermsEnum"
	}
	return "FuzzyTermsEnum"
}

// FloatToEdits 把「最小相似度」换算成编辑距离。
//
//	similarity >= 1f → min(similarity, 2)
//	similarity == 0f → 0（源码注释：0 means exact, not infinite # of edits!）
//	否则             → min((int)((1 - similarity) * termLen), 2)
func FloatToEdits(minimumSimilarity float32, termLen int) int {
	if minimumSimilarity >= 1.0 {
		return int(min(float64(minimumSimilarity), MaximumSupportedDistance))
	}
	if minimumSimilarity == 0.0 {
		return 0
	}
	return min(int(float64(1.0-minimumSimilarity)*float64(termLen)), MaximumSupportedDistance)
}

// ---------------------------------------------------- ParametricDescription

// ParametricDescription 参数化描述：状态数 / 接受态 / 位置。
type ParametricDescription struct {
	W          int
	N          int
	MinErrors  []int
}

// Size minErrors.length * (w + 1)
func (d *ParametricDescription) Size() int { return len(d.MinErrors) * (d.W + 1) }

// IsAccept 接受条件 `w - offset + minErrors[state] <= n`
func (d *ParametricDescription) IsAccept(absState int) bool {
	state := absState / (d.W + 1)
	offset := absState % (d.W + 1)
	if offset < 0 {
		panic("offset must be >= 0")
	}
	return d.W-offset+d.MinErrors[state] <= d.N
}

// GetPosition 最小边界函数：absState % (w + 1)
func (d *ParametricDescription) GetPosition(absState int) int {
	return absState % (d.W + 1)
}

// AcceptStates 所有接受态。
func (d *ParametricDescription) AcceptStates() []int {
	out := []int{}
	for s := 0; s < d.Size(); s++ {
		if d.IsAccept(s) {
			out = append(out, s)
		}
	}
	return out
}

// NewLev1 super(w, 1, new int[] {0, 1, 0, -1, -1})
func NewLev1(w int) *ParametricDescription {
	return &ParametricDescription{w, 1, []int{0, 1, 0, -1, -1}}
}

// NewLev1T super(w, 1, new int[] {0, 1, 0, -1, -1, -1})
func NewLev1T(w int) *ParametricDescription {
	return &ParametricDescription{w, 1, []int{0, 1, 0, -1, -1, -1}}
}

// NewLev2 super(w, 2, new int[] {0,1,2,0,1,-1,0,-1,0,-1,0,-1,-1,-1,-1,-2,
//
//	-1,-2,-1,-2,-1,-2,-2,-2,-2,-2,-2,-2,-2,-2})
func NewLev2(w int) *ParametricDescription {
	return &ParametricDescription{w, 2, []int{
		0, 1, 2, 0, 1, -1, 0, -1, 0, -1, 0, -1, -1, -1, -1, -2,
		-1, -2, -1, -2, -1, -2, -2, -2, -2, -2, -2, -2, -2, -2}}
}

// --------------------------------------------------------- LevenshteinAutomata

// LevenshteinAutomata 按源码顺序复刻构造过程。
type LevenshteinAutomata struct {
	Word        []int
	AlphaMax    int
	Alphabet    []int
	RangeLower  []int
	RangeUpper  []int
	NumRanges   int
	Descriptions [3]*ParametricDescription
}

// NewLevenshteinAutomata 抽字母表 → 算补集区间 → 选参数化描述。
//
// n=2 的 T 变体（Lev2TParametricDescription）本轮没取到源码，因此不做区分，
// n=2 一律用 Lev2 的骨架（状态数/接受判据只对非 T 变体严格成立）。
func NewLevenshteinAutomata(word []int, alphaMax int, withTranspositions bool) (*LevenshteinAutomata, error) {
	la := &LevenshteinAutomata{Word: word, AlphaMax: alphaMax}
	seen := map[int]bool{}
	for _, v := range word {
		if v > alphaMax {
			return nil, fmt.Errorf("alphaMax exceeded by symbol %d in word", v)
		}
		seen[v] = true
	}
	for v := range seen {
		la.Alphabet = append(la.Alphabet, v)
	}
	sort.Ints(la.Alphabet)

	lower := 0
	for _, higher := range la.Alphabet {
		if higher > lower {
			la.RangeLower = append(la.RangeLower, lower)
			la.RangeUpper = append(la.RangeUpper, higher-1)
			la.NumRanges++
		}
		lower = higher + 1
	}
	if lower <= alphaMax {
		la.RangeLower = append(la.RangeLower, lower)
		la.RangeUpper = append(la.RangeUpper, alphaMax)
		la.NumRanges++
	}

	w := len(word)
	lev1 := NewLev1(w)
	if withTranspositions {
		lev1 = NewLev1T(w)
	}
	la.Descriptions = [3]*ParametricDescription{nil, lev1, NewLev2(w)}
	return la, nil
}

// GetVector 特征向量 X(x, V)：从 pos 到 end 逐位左移，命中置 1。
func (la *LevenshteinAutomata) GetVector(x, pos, end int) int {
	vector := 0
	for i := pos; i < end; i++ {
		vector <<= 1
		if la.Word[i] == x {
			vector |= 1
		}
	}
	return vector
}

// VectorWindow toAutomaton 里用的窗口：end = pos + min(w - pos, 2n+1)。
func (la *LevenshteinAutomata) VectorWindow(x, pos, n int) (int, int) {
	end := pos + min(len(la.Word)-pos, 2*n+1)
	return la.GetVector(x, pos, end), end
}

// AutomatonPlan toAutomaton(n, prefix) 前半段能确定的量。
type AutomatonPlan struct {
	Kind           string
	Value          string
	Range          int
	NumStates      int
	NumTransitions int
	PrefixStates   int
	TotalStates    int
}

// ToAutomatonPlan n == 0 退化成 makeString(prefix + word)；n >= len(descriptions) 不支持。
func (la *LevenshteinAutomata) ToAutomatonPlan(n int, prefix string) AutomatonPlan {
	if n == 0 {
		s := ""
		for _, c := range la.Word {
			s += string(rune(c))
		}
		return AutomatonPlan{Kind: "makeString", Value: prefix + s}
	}
	if n >= len(la.Descriptions) {
		return AutomatonPlan{Kind: "unsupported"}
	}
	desc := la.Descriptions[n]
	return AutomatonPlan{
		Kind:           "dfa",
		Range:          2*n + 1,
		NumStates:      desc.Size(),
		NumTransitions: desc.Size() * min(1+2*n, len(la.Alphabet)),
		PrefixStates:   len([]rune(prefix)),
		TotalStates:    desc.Size() + len([]rune(prefix)),
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 07-数据存储/04-搜索引擎/模糊查询与编辑距离自动机/python/fuzzy.py
"""FuzzyQuery 参数语义与 Levenshtein 自动机的参数化描述。

忠实转写自 apache/lucene@main：
  lucene/core/src/java/org/apache/lucene/search/FuzzyQuery.java
  lucene/core/src/java/org/apache/lucene/util/automaton/LevenshteinAutomata.java
  .../automaton/Lev1ParametricDescription.java / Lev2ParametricDescription.java /
  Lev1TParametricDescription.java

**建模口径**：真正的 DFA 状态转移表（`toStates*` / `offsetIncrs*` 打包数组）这里不还原，
本 demo 断言的是**参数化描述的骨架** —— 状态数、接受态判据、位置函数、特征向量，
以及 FuzzyQuery 的参数校验与默认值。这些都能从源码逐行对上。
"""

MAXIMUM_SUPPORTED_DISTANCE = 2
CHARACTER_MAX_CODE_POINT = 0x10FFFF


# ------------------------------------------------------------------ FuzzyQuery

class FuzzyQuery(object):
    """FuzzyQuery 的构造参数与校验（源码四条 throw 全部复刻）。"""

    defaultMaxEdits = MAXIMUM_SUPPORTED_DISTANCE   # = 2
    defaultPrefixLength = 0
    defaultMaxExpansions = 50
    defaultTranspositions = True

    def __init__(self, term, max_edits=None, prefix_length=None,
                 max_expansions=None, transpositions=None):
        self.term = term
        self.max_edits = self.defaultMaxEdits if max_edits is None else max_edits
        self.prefix_length = self.defaultPrefixLength if prefix_length is None else prefix_length
        self.max_expansions = (self.defaultMaxExpansions if max_expansions is None
                               else max_expansions)
        self.transpositions = (self.defaultTranspositions if transpositions is None
                               else transpositions)

        if self.max_edits < 0 or self.max_edits > MAXIMUM_SUPPORTED_DISTANCE:
            raise ValueError("maxEdits must be between 0 and %d" % MAXIMUM_SUPPORTED_DISTANCE)
        if self.prefix_length < 0:
            raise ValueError("prefixLength cannot be negative.")
        if self.max_expansions <= 0:
            raise ValueError("maxExpansions must be positive.")

    @property
    def terms_enum_kind(self):
        """maxEdits == 0 时走 SingleTermsEnum（只能精确匹配），否则 FuzzyTermsEnum。"""
        return "SingleTermsEnum" if self.max_edits == 0 else "FuzzyTermsEnum"


def float_to_edits(minimum_similarity, term_len):
    """把「最小相似度」换算成编辑距离。三分支，源码照抄。

    * similarity >= 1f → min(similarity, 2)，即直接当整数编辑距离用
    * similarity == 0.0f → **0**（源码注释：0 means exact, not infinite # of edits!）
    * 否则 → min((int)((1 - similarity) * termLen), 2)
    """
    if minimum_similarity >= 1.0:
        return int(min(minimum_similarity, MAXIMUM_SUPPORTED_DISTANCE))
    if minimum_similarity == 0.0:
        return 0
    return min(int((1.0 - minimum_similarity) * term_len), MAXIMUM_SUPPORTED_DISTANCE)


# ---------------------------------------------------- ParametricDescription

class ParametricDescription(object):
    """参数化描述：状态数 / 接受态 / 位置 / 转移，全部以词长 w 为参数。"""

    def __init__(self, w, n, min_errors):
        self.w = w
        self.n = n
        self.min_errors = list(min_errors)

    def size(self):
        """minErrors.length * (w + 1)。"""
        return len(self.min_errors) * (self.w + 1)

    def is_accept(self, abs_state):
        """state = absState / (w+1)，offset = absState % (w+1)；
        接受条件 `w - offset + minErrors[state] <= n`。"""
        state = abs_state // (self.w + 1)
        offset = abs_state % (self.w + 1)
        assert offset >= 0
        return self.w - offset + self.min_errors[state] <= self.n

    def get_position(self, abs_state):
        """最小边界函数：absState % (w + 1)。"""
        return abs_state % (self.w + 1)

    def accept_states(self):
        return [s for s in range(self.size()) if self.is_accept(s)]


class Lev1ParametricDescription(ParametricDescription):
    def __init__(self, w):
        super(Lev1ParametricDescription, self).__init__(w, 1, [0, 1, 0, -1, -1])


class Lev1TParametricDescription(ParametricDescription):
    def __init__(self, w):
        super(Lev1TParametricDescription, self).__init__(w, 1, [0, 1, 0, -1, -1, -1])


class Lev2ParametricDescription(ParametricDescription):
    def __init__(self, w):
        super(Lev2ParametricDescription, self).__init__(w, 2, [
            0, 1, 2, 0, 1, -1, 0, -1, 0, -1, 0, -1, -1, -1, -1, -2,
            -1, -2, -1, -2, -1, -2, -2, -2, -2, -2, -2, -2, -2, -2])


# --------------------------------------------------------- LevenshteinAutomata

class LevenshteinAutomata(object):
    """按源码顺序复刻构造过程：抽字母表 → 算补集区间 → 选参数化描述。"""

    def __init__(self, word, alpha_max=CHARACTER_MAX_CODE_POINT,
                 with_transpositions=True):
        self.word = list(word)
        self.alpha_max = alpha_max

        seen = set()
        for v in self.word:
            if v > alpha_max:
                raise ValueError("alphaMax exceeded by symbol %d in word" % v)
            seen.add(v)
        self.alphabet = sorted(seen)

        # 字母表之外的 Unicode 区间（特征向量恒为 0）
        self.range_lower = []
        self.range_upper = []
        num_ranges = 0
        lower = 0
        for higher in self.alphabet:
            if higher > lower:
                self.range_lower.append(lower)
                self.range_upper.append(higher - 1)
                num_ranges += 1
            lower = higher + 1
        if lower <= alpha_max:
            self.range_lower.append(lower)
            self.range_upper.append(alpha_max)
            num_ranges += 1
        self.num_ranges = num_ranges

        w = len(self.word)
        # n=2 的 T 变体（Lev2TParametricDescription）本轮没取到源码，
        # 因此这里**不做区分**：n=2 一律用 Lev2ParametricDescription 的骨架，
        # 并在 README 标注该限制（状态数/接受判据只对非 T 变体严格成立）。
        self.descriptions = [
            None,                                           # n=0 不需要参数化描述
            Lev1TParametricDescription(w) if with_transpositions
            else Lev1ParametricDescription(w),
            Lev2ParametricDescription(w),
        ]

    def to_automaton_plan(self, n, prefix=""):
        """返回构造 DFA 前能确定的几个量；对应 toAutomaton(n, prefix) 的前半段。

        n == 0        → 退化成 makeString(prefix + word)
        n >= len(desc) → 源码直接 return null（不支持）
        """
        if n == 0:
            return {"kind": "makeString", "value": prefix + "".join(
                chr(c) for c in self.word)}
        if n >= len(self.descriptions):
            return {"kind": "unsupported"}
        desc = self.descriptions[n]
        num_states = desc.size()
        return {
            "kind": "dfa",
            "range": 2 * n + 1,
            "numStates": num_states,
            "numTransitions": num_states * min(1 + 2 * n, len(self.alphabet)),
            "prefixStates": len(prefix),
            "totalStates": num_states + len(prefix),
        }

    def get_vector(self, x, pos, end):
        """特征向量 X(x, V)：从 pos 到 end 逐位左移，命中置 1。"""
        vector = 0
        for i in range(pos, end):
            vector <<= 1
            if self.word[i] == x:
                vector |= 1
        return vector

    def vector_window(self, x, pos, n):
        """toAutomaton 里用的窗口：end = pos + min(w - pos, 2n+1)。"""
        end = pos + min(len(self.word) - pos, 2 * n + 1)
        return self.get_vector(x, pos, end), end
